### PR TITLE
MAINT: switch to using spmatrix.toarray instead of .todense

### PR DESCRIPTION
**Pull request by akhmerov**
#### Reference issue

First step towards resolving #14494

#### What does this implement/fix?

This PR stops using the `spmatrix.todense` in favor of `spmatrix.toarray` because the former produces a deprecated np.matrix. The replacement `.todense` → `.toarray` is only done where reasonable, specifically:
- In tests, unless the tests aim to check the output type.
- In math operations when combining with a numpy array (otherwise array + sparse matrix = dense matrix)
- In docstrings

Additionally, it eliminates most usage of np.matrix from sparse tests, unless again the type is important. This is tangential to the main PR goal, but it's conceptually related.

#### Additional information

The sparse library is not cleanly formatted for historical reasons. Following previous advice by @tupui, I fixed the formatting only in the lines affected by my work.

### CODE REVIEW

**Comment by akhmerov** [comment]
The failing test is due to a timeout, and is unrelated to my changes.

File: scipy/sparse/base.py [hunk 1]
```diff
@@ -396,11 +396,11 @@ def _sub_sparse(self, other):
         return self.tocsr()._sub_sparse(other)
 
     def _sub_dense(self, other):
-        return self.todense() - other
+        return self.toarray() - other
```
Comment (perimosocordiae) [inline scipy/sparse/base.py:399]: This and the `_rsub_dense` change aren't backwards compatible, unfortunately. Let's leave these out of this PR.

Comment (perimosocordiae) [reply scipy/sparse/base.py:399]: And the `_divide` changes as well.

File: scipy/sparse/compressed.py [hunk 2]
```diff
@@ -229,7 +229,7 @@ def __eq__(self, other):
                 return self._scalar_binopt(other, operator.eq)
         # Dense other.
         elif isdense(other):
-            return self.todense() == other
+            return self.toarray() == other
```
Comment (perimosocordiae) [inline scipy/sparse/compressed.py:232]: All the changes in this file break back-compat.

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py [hunk 3]
```diff
@@ -57,13 +57,15 @@ def test_laplacian_value_error():
...
-            'sparse.diags([1, 1], [-1, 1], shape=(4,4)).todense()',
-            'np.asarray(sparse.diags([1, 1], [-1, 1], shape=(4,4)).todense())',
-            'np.vander(np.arange(4)) + np.vander(np.arange(4)).T')
+    symmetric_mats = (
+        'np.arange(10) * np.arange(10)[:, np.newaxis]',
+        'np.ones((7, 7))',
+        'np.eye(19)',
+        'sparse.diags([1, 1], [-1, 1], shape=(4, 4))',
+        'sparse.diags([1, 1], [-1, 1], shape=(4, 4)).toarray()',
+        'np.asarray(sparse.diags([1, 1], [-1, 1], shape=(4, 4)).toarray())',
```
Comment (perimosocordiae) [inline scipy/sparse/csgraph/tests/test_graph_laplacian.py:66]: this `np.asarray()` call is redundant now

File: scipy/sparse/dok.py [hunk 4]
```diff
@@ -267,7 +267,7 @@ def __add__(self, other):
             csc = self.tocsc()
             new = csc + other
         elif isdense(other):
-            new = self.todense() + other
+            new = self.toarray() + other
```
Comment (perimosocordiae) [inline scipy/sparse/dok.py:270]: Changes in this file break back-compat

File: scipy/sparse/tests/test_sparsetools.py [hunk 5]
```diff
@@ -53,17 +53,17 @@ def test_regression_std_vector_dtypes():
     # Regression test for gh-3780, checking the std::vector typemaps
     # in sparsetools.cxx are complete.
     for dtype in supported_dtypes:
-        ad = matrix([[1, 2], [3, 4]]).astype(dtype)
+        ad = np.array([[1, 2], [3, 4]]).astype(dtype)
         a = csr_matrix(ad, dtype=dtype)
 
         # getcol is one function using std::vector typemaps, and should not fail
-        assert_equal(a.getcol(0).todense(), ad[:,0])
+        assert_equal(a.getcol(0).toarray(), ad[:, [0]])
```
Comment (perimosocordiae) [inline scipy/sparse/tests/test_sparsetools.py:60]: alternate slicing `ad[:, :1]` looks nicer to me

File: scipy/sparse/tests/test_sparsetools.py [hunk 6]
```diff
@@ -53,17 +53,17 @@ def test_regression_std_vector_dtypes():
...
 
 
 @pytest.mark.slow
 @pytest.mark.xfail_on_32bit("Can't create large array for test")
 def test_nnz_overflow():
-    # Regression test for gh-7230 / gh-7871, checking that coo_todense
+    # Regression test for gh-7230 / gh-7871, checking that coo_toarray
```
Comment (perimosocordiae) [inline scipy/sparse/tests/test_sparsetools.py:66]: This may not be an intentional change?

**Review by perimosocordiae — CHANGES_REQUESTED** [review]
First: my apologies for the long delay in reviewing, and thanks for your efforts to clean up the scipy.sparse codebase!

I'm happy with the docstring and test changes here, and I agree we should prefer `.toarray()` over `.todense()` in most cases. I don't think we should include the various changes to binary operator methods, though, as these will result in arrays being returned where matrices used to be. This is a recipe for sad users, as the ndarray and matrix APIs are just different enough to produce some really nasty bugs if an array appears where a matrix was expected (or vice versa).

**Comment by akhmerov** [comment]
> I don't think we should include the various changes to binary operator methods, though, as these will result in arrays being returned where matrices used to be. This is a recipe for sad users, as the ndarray and matrix APIs are just different enough to produce some really nasty bugs if an array appears where a matrix was expected (or vice versa).

@perimosocordiae I'm OK with reverting the changes, but please confirm that it's the preference. The reason for me asking is that as things are *right now*, silently casting arrays to matrices is an ongoing recipe for sad users of scipy.sparse, see https://github.com/scipy/scipy/pull/14488#discussion_r678451919

While I have no way of telling this, I expect that most users avoid np.matrix API, and cast `array + spmatrix` into array immediately after performing the operation.

**Comment by WarrenWeckesser** [comment]
@perimosocordiae is correct, we can't change the return type of the methods that currently return `matrix` to `ndarray`.   For better or worse, SciPy's sparse matrices have always acted like sparse versions of the `matrix` class.  We will have to figure out how to make the transition away from NumPy's `matrix` eventually, but a backwards-incompatible change like this is not the way to do it.

**Comment by WarrenWeckesser** [comment]
@akhmerov, what I should have said *first* is: *thank you* for the pull request!  That's a lot of work, and much of will be useful, but the complete switch away from `matrix` is too much at the moment.

File: scipy/sparse/tests/test_sparsetools.py (hunk 6 above)
Comment (akhmerov) [reply scipy/sparse/tests/test_sparsetools.py:66]: The code below didn't invoke `todense` also before the change (line 79), so this is a fix of the comment, and it's intentional.

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py (hunk 3 above)
Comment (akhmerov) [reply scipy/sparse/csgraph/tests/test_graph_laplacian.py]: This change was actually against the original intention, which checked Laplacian with np.matrix input. I have fixed this now.

**Comment by akhmerov** [comment]
I've reverted the changes that break backwards compatibility. Related to that, I would appreciate some feedback on #14494 and/or my [proposal](https://mail.python.org/pipermail/scipy-dev/2021-August/024959.html) on the mailing list.

Also the circleCI failure seems unrelated.

**Review by perimosocordiae — APPROVED** [review]
+1 to merge, thanks for being patient with me.

I left one comment on the massive test_base.py file, but otherwise everything looks fine.

**Pull request merged**